python/ray/serve/backend_state.py

```diff
@@ -539,6 +539,9 @@ def __init__(self, controller_name: str, detached: bool,
         self._goal_manager = goal_manager
         self._replicas: Dict[BackendTag, ReplicaStateContainer] = dict()
         self._backend_metadata: Dict[BackendTag, BackendInfo] = dict()
+        # TODO(architkulkarni): Limit the size of this, currently it grows
+        # without bound as more and more deployments are deleted.
+        self._deleted_deployments: Dict[BackendTag, BackendInfo] = dict()
```

Comment:
let's at least bound this with ordereddict? https://docs.python.org/3/library/collections.html#collections.OrderedDict
```python
d = OrderedDict()
if len(d) >= 100:
    d.popitem(last=False)
d[k] = v
```

Reply:
Can we call it `deleted_backend_metadata` instead since it's exactly the same type and content as `_backend_metadata` with same key value for assignment, so we don't have to introduce another term like "deployment" ?

Also, what happens if the same backend tag (for example, "/endpoint") got deployed and deleted multiple times, do we intend to always keep the last one or rather keep history of deployments for the same endpoint overtime ? From my local script 

```
import ray
from ray import serve

ray.init()
serve.start()

@serve.deployment(num_replicas=1)
class ConstructorFailureDeployment:
    def __init__(self):
        raise RuntimeError("Intentionally throwing ")

    async def serve(self, request):
        return "hi"

ConstructorFailureDeployment.deploy()
```


Our backend state thinks backend tag is "ConstructorFailureDeployment" which seems to just be the class name without any randomized suffix, thus the chance of reusing the same backend tag could be high.

Reply:
`deleted_backend_metadata` name change sounds good, I'll change it.  

Regarding repeated deletes and deploys of a given deployment, just discussed this on Slack with @kombuchafox and for now the current behavior (always keep the last one) is okay, the consumer of the snapshot will keep track of the historical data.